core/CronArchive.php

```diff
@@ -365,19 +365,36 @@ public function run()
         $this->logSection("START");
         $this->logger->info("Starting Matomo reports archiving...");
 
-        do {
+        $numWebsitesScheduled = $this->websites->getNumSites();
+        $numWebsitesArchived = 0;
 
+        do {
             if ($this->isMaintenanceModeEnabled()) {
                 $this->logger->info("Archiving will stop now because maintenance mode is enabled");
                 return;
             }
 
             $idSite = $this->websites->getNextSiteId();
+            $numWebsitesArchived++;
```

Comment:
Please note this does not 100% fix it. For example, say there are 50 sites in the initial list that are left to be archived. Now there are 2 archivers activate at the same time and each archive 25 sites. This means `$numWebsitesArchived = 25`, `$numWebsitesScheduled = 50` and even though the `sharedArchiveSiteId` has been finished and will be reseted, we will still run for 25 more sites than it should be. Not sure how to avoid this. Was thinking of setting

`$numWebsitesScheduled = min($numWebsitesScheduled, count($this->websites-> getAllSiteIdsToArchive()))` but then we would stop processing too early.

I suppose what would help is like:
```
$numWebsitesLeft = count($this->websites-> getAllSiteIdsToArchive());
if ($numWebsitesLeft < $numWebsitesScheduled) {
   $numWebsitesScheduled = $numWebsitesLeft;
   $numWebsitesArchived = 1;
}
```

that way, if the number of remaining sites was reduced and not reseted, it may "auto correct itself but not 100% if that works?

Reply:
> $numWebsitesScheduled = $numWebsitesLeft; might make it process more sites, right? 

nah shouldn't because it would be only updated when there are less sites left unless I don't see it right.

> Oh, here's another idea that may or may not work, what if there was a special 'null' entry in queue that would signal that this batch was done and this core:archive should stop? I guess that would also stop all future jobs too...

yep exactly.

> Or, for yet another random idea, what if the SharedSiteIds option had the current hour in the option name, so the once the core:archive instances finish that specific queue, they quit?

that's not quite working either because you might not run an archive command only once per hour but even more often. And sometimes the first sites take very long in which we need to make sure another job an hour later starts working on the next idsite.

This here is currently not a huge problem and we wouldn't even need to add another fix to it at all. But be good to let `core:archive` tasks finish rather sooner than later.

Reply:
> nah shouldn't because it would be only updated when there are less sites left unless I don't see it right.

it would be updated when there are less than the original (ie, 50), but not less than the current remaining jobs at the end (ie, 25). although it happens on every iteration, right? that could get weird if jobs keep getting added...

Reply:
fyi @mattab @diosmosis I added another patch to this PR to detect this, and I implemented the patch in `SharedSiteIds` to on top detect when the shared site IDs have been processed.